File: src/python/minnow_comms/minnow_app_threaded.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-

import sys
import zmq
import time
import signal
import yaml
from threading import Thread, Event, Lock

class Subscriber(Thread):
    def __init__(self, app, address, topic, callback):
        super().__init__()
        self.app_ = app
        self.address_ = address
        self.topic_ = topic
        self.callback_ = callback
        self.mutex_ = Lock()
        self.new_msg_ = False
        self.msg_ = None
        self.zmq_context_ = app.zmq_context_
        self.receive_active_ = Event()

    def run(self):
        print("[T] Starting subscriber thread for topic: {}".format(self.topic_))
        subscriber = self.zmq_context_.socket(zmq.SUB)
        subscriber.connect(self.address_)
        subscriber.setsockopt_string(zmq.SUBSCRIBE, self.topic_)
        poller = zmq.Poller()
        poller.register(subscriber, zmq.POLLIN)
        while not self.receive_active_.is_set():
            evts = poller.poll(1000)
            if evts:
                message = subscriber.recv()

                topic_msg = message.split(b'_',1)
                topic = topic_msg[0].decode('UTF-8')
                if topic == self.topic_:
                    self.mutex_.acquire()
                    self.new_msg_ = True
                    self.msg_ = message
                    self.mutex_.release()
                else:
                    subscribed = self.app_.check_subscription_to_topic(topic)
                    if not subscribed:
                        self.app_.subscribe(topic, self.callback_)

        print("[T] Stopping subscriber thread for topic: {}".format(self.topic_))

class App:
    def __init__(self):
        self.zmq_context_ = zmq.Context()
        self.zmq_socket_ = self.zmq_context_.socket(zmq.PUB)
        self.mutex_ = Lock()
        self.subscriptions_ = []
        self.subscription_topics_ = []
        self.sleep_us_ = 10000  # 100Hz default tick

        self.name_ = None
        self.config_ = None
        self.protocol_ = None
        self.host_ = None
        self.port_ = None
        self.subscriber_address_ = None
        self.loop_start_ = None
        self.loop_end_ = None
        self.loop_us_ = None

        signal.signal(signal.SIGINT, self.exit_signal)

    def exit_signal(self, sig, frame):
        for subscriber in self.subscriptions_:
            subscriber.receive_active_.set()
        time.sleep(1)
        for subscriber in self.subscriptions_:
            print("[M] Subscription for {} killed.".format(subscriber.topic_))
            subscriber.join()
        time.sleep(1.5)
        print("Exiting Minnow App...")
        sys.exit(0)

    def subscribe(self, topic, callback):
        print("[M] Subscription for {} created.".format(topic))
        subscriber = Subscriber(self, self.subscriber_address_, topic, callback)
        self.mutex_.acquire()
        self.subscriptions_.append(subscriber)
        self.subscription_topics_.append(topic)
        self.mutex_.release()
        subscriber.start()

    def check_subscription_to_topic(self, topic):
        subscribed = False
        for subscribed_topic in self.subscription_topics_:
            if subscribed_topic == topic:
                subscribed = True
                break
        return subscribed

    def check_subscriptions(self):
        self.mutex_.acquire()
        for subscriber in self.subscriptions_:
            if subscriber.new_msg_:
                print("[M] New message from {}.".format(subscriber.topic_))
                subscriber.mutex_.acquire()
                subscriber.new_msg_ = False
                topic_msg = subscriber.msg_.split(b'_',1)
                subscriber.mutex_.release()
                topic = topic_msg[0].decode('UTF-8')
                msg = topic_msg[1]
                subscriber.callback_(msg, topic)
        self.mutex_.release()

    def publish_string(self, topic, msg):
        self.zmq_socket_.send_string(topic + '_' + msg)

    def publish(self, topic, msg):
        self.zmq_socket_.send(topic.encode('UTF-8') + b'_' + msg)

    def set_name(self, name):
        self.name_ = name

    def set_config(self, config_file):
        try:
            with open(config_file, 'r') as file:
                self.config_ = yaml.load(file)
        except:
            print("[C] Configuration file \"{}\" not found... Exiting.".format(config_file))
            self.exit_signal(None, None)

        try:
            self.protocol_ = self.config_["protocol"]
            self.host_ = self.config_["host"]
            if self.protocol_ == "ipc":
                print("[C] Using inter-process communications (ipc).")
            elif self.protocol_ == "tcp":
                print("[C] Using transmission control protocol (tcp).")
                self.port_ = int(self.config_["port"])
            else:
                print("[C] Unknown communications protocol... Exiting.")
                self.exit_signal(None, None)
        except:
            print("[C] Configuration file must define \"protocol\", \"host\" and, if required, \"port\"... Exiting.")
            self.exit_signal(None, None)

    def get_config_parameter(self, type, config_param):
        try:
            val = self.config_[self.name_][config_param]
        except:
            print("[C] Configuration parameter \"{}\" for Minnow App \"{}\" not found... Exiting.".format(config_param, self.name_))
            self.exit_signal(None, None)
        try:
            val = type(val)
        except:
            print("[C] Configuration parameter \"{}\" for Minnow App \"{}\" could not be cast into type \"{}\"... Exiting.".format(config_param, self.name_, type))
            self.exit_signal(None, None)
        return val

    def set_hz(self, tick_hz):
        if tick_hz < 0:
            print("[C] Configuration parameter \"tick\" for Minnow App \"{}\" must be positive or zero... Exiting.".format(self.name_))
        elif tick_hz == 0:
            self.sleep_us_ = 0
        else:
            self.sleep_us_ = int(1e6/tick_hz)

    def init(self): # must be implemented by subclasses
        for subscriber in self.subscriptions_:
            subscriber.receive_active_.set()
        time.sleep(1)
        for subscriber in self.subscriptions_:
            print("[M] Subscription for {} killed.".format(subscriber.topic_))
            subscriber.join()
        time.sleep(1.5)
        print("Exiting Minnow App...")
        raise NotImplementedError('Inheriting apps must implement init() method!')

    def process(self): # must be implemented by subclasses
        for subscriber in self.subscriptions_:
            subscriber.receive_active_.set()
        time.sleep(1)
        for subscriber in self.subscriptions_:
            print("[M] Subscription for {} killed.".format(subscriber.topic_))
            subscriber.join()
        time.sleep(1.5)
        print("Exiting Minnow App...")
        raise NotImplementedError('Inheriting apps must implement process() method!')

    def run(self):
        print("[M] Starting Minnow App \"{}\"".format(self.name_))

        if self.protocol_ == "ipc":
            socket = "ipc://" + self.host_ + ".sub"
        elif self.protocol_ == "tcp":
            socket = "tcp://" + self.host_ + ":" + self.port_+1
        print("[C] Publishing address: {}".format(socket))
        self.zmq_socket_.connect(socket)
        if self.protocol_ == "ipc":
            self.subscriber_address_ = "ipc://" + self.host_ + ".pub"
        elif self.protocol_ == "tcp":
            self.subscriber_address_ = "tcp://" + self.host_ + ":" + self.port_
        print("[C] Subscribing address: {}".format(self.subscriber_address_))

        self.init()

        while True:
            if self.sleep_us_ != 0:
                self.loop_start_ = time.time()

            self.check_subscriptions()
            self.process()

            if self.sleep_us_ != 0:
                self.loop_end_ = time.time()
                self.loop_us_ = 1e6*(self.loop_end_ - self.loop_start_)
                rem_usecs = self.sleep_us_ - self.loop_us_
                if rem_usecs > 0:
                    time.sleep(rem_usecs/1e6)

File: src/cpp/sandbox/pub_sub_testing/minnow_app_threaded.cpp
#include <iostream>
#include <sstream>
#include <signal.h>
#include <zmq.hpp>
#include "minnow_app_threaded.h"

using namespace std::chrono;

Subscriber::Subscriber(App* app, std::string address, std::string topic, std::function<void(uint8_t* msg, size_t msg_size, std::string topic)> callback) : receive_active_(false), new_msg_(false) {
  app_ = app;
  zmq_context_ = app_->zmq_context_;
  address_ = address;
  topic_ = topic;
  callback_ = callback;
}

Subscriber::~Subscriber() {
  receive_thread_.join();
}

void Subscriber::Start() {
  receive_active_ = true;
  receive_thread_ = std::thread(&Subscriber::Run, this);
}

void Subscriber::Stop() {
  receive_active_ = false;
}

void Subscriber::Run() {
  std::stringstream ss;
  ss << "[T] Starting subscriber thread for topic: " << topic_;
  Print(ss.str());
  ss.str("");

  // variables for message matching
  std::string msg_str;
  size_t msg_size;
  bool subscribed;

  zmq::socket_t subscriber(*zmq_context_, ZMQ_SUB);
  zmq_connect(subscriber, address_.c_str());
  zmq_setsockopt(subscriber, ZMQ_SUBSCRIBE, topic_.c_str(), topic_.size());
  std::vector<zmq::pollitem_t> p = {{subscriber, 0, ZMQ_POLLIN, 0}};
  while(receive_active_) {
    zmq_poll(p.data(), 1, 1000);
    if (p[0].revents & ZMQ_POLLIN) {
      mutex_.lock();
      subscriber.recv(&msg_, ZMQ_DONTWAIT);
      // copy message into local variable for checking
      msg_str.assign(static_cast<char *>(msg_.data()), msg_.size());
      msg_size = msg_.size();
      mutex_.unlock();

      // check for message match - if not a match, spin up a new subscriber for the new message
      size_t msg_start = msg_str.find_first_of("_");
      std::string topic = msg_str.substr(0,msg_start);
      if(topic == topic_) {
        new_msg_ = true;
      } else {
        new_msg_ = false;
        subscribed = app_->CheckSubscriptionToTopic(topic);
        if(!subscribed) {
          app_->Subscribe(topic, callback_);
        }
      }
    }
  }

  ss << "[T] Stopping subscriber thread for topic: " << topic_;
  Print(ss.str());
}

////////////////////////////////////////////////////////////////////////////////

App::App() {
  zmq_context_ = new zmq::context_t(1);
  zmq_socket_ = new zmq::socket_t(*zmq_context_, ZMQ_PUB);
  sleep_us_ = 10000;  // 100Hz default

  cpphandler = std::bind(&App::ExitSignal, this, std::placeholders::_1);
  struct sigaction sigIntHandler;
  sigIntHandler.sa_handler = signal_handler;
  sigemptyset(&sigIntHandler.sa_mask);
  sigIntHandler.sa_flags = 0;
  sigaction(SIGINT, &sigIntHandler, NULL);
}

App::~App() {
  for(std::vector<Subscriber*>::iterator it = subscriptions_.begin(); it != subscriptions_.end(); ++it) {
    delete *it;
  }
  delete zmq_socket_;
  delete zmq_context_;
}

void App::ExitSignal(int s) {
  std::stringstream ss;
  for(std::vector<Subscriber*>::iterator it = subscriptions_.begin(); it != subscriptions_.end(); ++it) {
    ss << "[M] Subscription for " << (*it)->topic_ << " killed.";
    Print(ss.str());
    ss.str("");

    (*it)->Stop();
  }
  std::this_thread::sleep_for(milliseconds(1500));
  Print("Exiting Minnow App...");
  zmq_ctx_destroy(zmq_context_);
  exit(1);
}

void App::Subscribe(std::string topic, std::function<void(uint8_t* msg, size_t msg_size, std::string topic)> callback) {
  std::stringstream ss;
  ss << "[M] Subscription for " << topic << " created.";
  Print(ss.str());

  Subscriber* sub = new Subscriber(this, subscriber_address_, topic, callback);
  mutex_.lock();
  subscriptions_.push_back(sub);
  subscription_topics_.push_back(topic);
  mutex_.unlock();
  sub->Start();
}

bool App::CheckSubscriptionToTopic(std::string topic) {
  bool subscribed = false;
  for(std::vector<std::string>::iterator it = subscription_topics_.begin(); it != subscription_topics_.end(); ++it) {
    if(topic == (*it)) subscribed = true;
  }
  return subscribed;
}

void App::CheckSubscriptions() {
  std::stringstream ss;
  mutex_.lock();
  for(std::vector<Subscriber*>::iterator it = subscriptions_.begin(); it != subscriptions_.end(); ++it) {
    if((*it)->new_msg_) {
      ss << "[M] New message from " << (*it)->topic_ << ".";
      Print(ss.str());
      ss.str("");
      (*it)->new_msg_ = false;
      std::string msg_str;
      size_t msg_size;
      (*it)->mutex_.lock();
      msg_str.assign(static_cast<char *>((*it)->msg_.data()), (*it)->msg_.size());
      msg_size = (*it)->msg_.size();
      (*it)->mutex_.unlock();
      size_t msg_start = msg_str.find_first_of("_");
      (*it)->callback_((uint8_t*)msg_str.data()+(msg_start+1), msg_size-(msg_start+1), msg_str.substr(0,msg_start));
    }
  }
  mutex_.unlock();
}

void App::PublishString(std::string topic, std::string msg, size_t msg_size) {
  Publish(topic, (uint8_t*)msg.data(), msg.length());
}

void App::Publish(std::string topic, uint8_t* msg, size_t msg_size) {
  topic.append("_");
  unsigned int total_len = msg_size + topic.length();
  uint8_t *buffer= new uint8_t[total_len];
  memcpy(buffer, topic.data(), topic.length());
  memcpy(buffer+topic.length(), msg, msg_size);
  int rc = zmq_send(*zmq_socket_, buffer, (size_t)total_len, 0);
}

void App::SetConfig(std::string config_file) {
  try {
		config_ = YAML::LoadFile(config_file);
	} catch (...) {
		std::cout << "[C] Configuration file \"" << config_file << "\" not found... Exiting." << std::endl;
		exit(0);
	}
  try {
    protocol_ = config_["protocol"].as<std::string>();
    host_ = config_["host"].as<std::string>();
    if(protocol_ == "ipc") {
      std::cout << "[C] Using inter-process communications (ipc)." << std::endl;
    } else if(protocol_ == "tcp") {
      std::cout << "[C] Using transmission control protocol (tcp)." << std::endl;
      port_ = config_["port"].as<int>();
    } else {
      std::cout << "[C] Unknown communications protocol... Exiting." << std::endl;
      exit(0);
    }
  } catch (...) {
    std::cout << "[C] Configuration file must define \"protocol\", \"host\" and, if required, \"port\"... Exiting." << std::endl;
    exit(0);
  }
}

void App::Run() {
  std::stringstream ss;
  ss << "[M] Starting Minnow App \"" << name_ << "\"!";
  Print(ss.str());
  ss.str("");

  if(protocol_ == "ipc") {
    ss << "ipc://" << host_ << ".sub";
  } else if(protocol_ == "tcp") {
    ss << "tcp://" << host_ << ":" << port_+1;
  }
  std::string socket = ss.str();
  std::cout << "[C] Publishing address: " << socket << std::endl;
  zmq_connect(*zmq_socket_, socket.c_str());
  ss.str("");
  if(protocol_ == "ipc") {
    ss << "ipc://" << host_ << ".pub";
  } else if(protocol_ == "tcp") {
    ss << "tcp://" << host_ << ":" << port_;
  }
  subscriber_address_ = ss.str();
  std::cout << "[C] Subscribing address: " << subscriber_address_ << std::endl;

  Init();

  while(true) {
    if(sleep_us_ != 0) loop_start_ = high_resolution_clock::now();

    CheckSubscriptions();
    Process();

    if(sleep_us_ != 0) {
      loop_end_ = high_resolution_clock::now();
      loop_us_ = duration_cast<microseconds>(loop_end_ - loop_start_);
      int rem_usecs = sleep_us_ - loop_us_.count();
      if(rem_usecs > 0) {
        std::this_thread::sleep_for(microseconds(rem_usecs));
      }
    }
  }
}
